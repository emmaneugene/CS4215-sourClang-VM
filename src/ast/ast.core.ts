import { Registers } from '../typings/microcode'
import { Declaration, FunctionDefinition } from './ast.declaration'

/**
 * Defines the supported datatypes.
 */
export enum DataType {
  VOID = 'void',
  CHAR = 'char',
  UNSIGNED_CHAR = 'unsigned_char',
  SHORT = 'short',
  UNSIGNED_SHORT = 'unsigned_short',
  INT = 'int',
  UNSIGNED_INT = 'unsigned_int',
  LONG = 'long',
  UNSIGNED_LONG = 'unsigned_long',
  FLOAT = 'float',
  DOUBLE = 'double',

  STRUCT = 'struct',

  FUNCTION = 'function',

  /**
   * Unknown during parsing.
   * E.g. int z = x;
   */
  UNKNOWN = 'unknown'
}

/**
 * Describes the position of the code.
 */
export interface Position {
  /** >= 1 */
  line: number
  /** >= 0 */
  column: number
}

/**
 * Describes the position of the code.
 */
export interface SourceLocation {
  source?: string | undefined
  start: Position
  end: Position
}

/**
 * A base node to describe the
 * source of the code.
 */
export interface BaseNode {
  type: string
  source?: string | undefined
  loc?: SourceLocation | undefined
  range?: [number, number] | undefined
}

/**
 * A JS object to describe a struct definition.
 */
export type StructDef = { [attribute: string]: StructDef | TypeList }

/**
 * Used to describe a type of an identifier.
 */
export type TypeList = {
  typeList: Array<'*' | DataType>
  structDef?: StructDef | undefined
  functionParams?: TypeList[]
  functionHasVariableArguments?: boolean
}

/**
 * The root node of an AST.
 */
export interface Program extends BaseNode {
  type: 'Program'
  body: Array<Declaration | FunctionDefinition>
}

/**
 * Used to describe an array of an identifier.
 */
export type Address =
  | {
      isInstructionAddr: true
    }
  | {
      isInstructionAddr: false
      address: [Registers, number]
    }
