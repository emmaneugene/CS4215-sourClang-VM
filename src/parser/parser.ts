import { CharStreams, CommonTokenStream } from 'antlr4ts'
import * as es from 'estree'

import { SourCLexer } from '../lang/SourCLexer'
import { Context, ErrorSeverity } from '../types'
import { SourCParser, TranslationUnitContext } from './../lang/SourCParser'
import { FatalSyntaxError } from './error'
import StatementGenerator from './statementGenerator'

function convertStatement(translationUnit: TranslationUnitContext): es.Statement {
  const generator = new StatementGenerator()
  return translationUnit.accept(generator)
}

function convertSource(translationUnit: TranslationUnitContext): es.Program {
  return {
    type: 'Program',
    sourceType: 'script',
    body: [convertStatement(translationUnit)]
  }
}

export function parse(source: string, context: Context) {
  let program: es.Program | undefined

  if (context.variant === 'calc') {
    const inputStream = CharStreams.fromString(source)
    const lexer = new SourCLexer(inputStream)
    const tokenStream = new CommonTokenStream(lexer)
    const parser = new SourCParser(tokenStream)
    parser.buildParseTree = true
    try {
      const tree = parser.translationUnit()
      program = convertSource(tree)
    } catch (error) {
      if (error instanceof FatalSyntaxError) {
        context.errors.push(error)
      } else {
        throw error
      }
    }
    const hasErrors = context.errors.find(m => m.severity === ErrorSeverity.ERROR)
    if (program && !hasErrors) {
      console.log(JSON.stringify(program, null, 2))
      return program
    } else {
      return undefined
    }
  } else {
    return undefined
  }
}

/*
import { CalcLexer } from '../lang/CalcLexer'
import {
  AdditionContext,
  CalcParser,
  DivisionContext,
  ExpressionContext,
  MultiplicationContext,
  NumberContext,
  ParenthesesContext,
  PowerContext,
  StartContext,
  SubtractionContext
} from '../lang/CalcParser'
import { CalcVisitor } from '../lang/CalcVisitor'

class ExpressionGenerator implements CalcVisitor<es.Expression> {
  visitNumber(ctx: NumberContext): es.Expression {
    return {
      type: 'Literal',
      value: parseInt(ctx.text),
      raw: ctx.text,
      loc: contextToLocation(ctx)
    }
  }
  visitParentheses(ctx: ParenthesesContext): es.Expression {
    return this.visit(ctx.expression())
  }
  visitPower(ctx: PowerContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '^',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitMultiplication(ctx: MultiplicationContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '*',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }
  visitDivision(ctx: DivisionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '/',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }
  visitAddition(ctx: AdditionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '+',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitSubtraction(ctx: SubtractionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '-',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitExpression?: ((ctx: ExpressionContext) => es.Expression) | undefined
  visitStart?: ((ctx: StartContext) => es.Expression) | undefined

  visit(tree: ParseTree): es.Expression {
    return tree.accept(this)
  }
  visitChildren(node: RuleNode): es.Expression {
    const expressions: es.Expression[] = []
    for (let i = 0; i < node.childCount; i++) {
      expressions.push(node.getChild(i).accept(this))
    }
    return {
      type: 'SequenceExpression',
      expressions
    }
  }
  visitTerminal(node: TerminalNode): es.Expression {
    return node.accept(this)
  }

  visitErrorNode(node: ErrorNode): es.Expression {
    throw new FatalSyntaxError(
      {
        start: {
          line: node.symbol.line,
          column: node.symbol.charPositionInLine
        },
        end: {
          line: node.symbol.line,
          column: node.symbol.charPositionInLine + 1
        }
      },
      `invalid syntax ${node.text}`
    )
  }
}

function convertExpression(expression: ExpressionContext): es.Expression {
  const generator = new ExpressionGenerator()
  return expression.accept(generator)
}
*/
