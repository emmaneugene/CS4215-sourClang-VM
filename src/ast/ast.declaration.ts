import { Address, BaseNode, TypeList } from './ast.core'
import { Expression, Identifier, SequenceExpression } from './ast.expression'
import { Statement } from './ast.statement'

/**
 * A base interface to describe declarations.
 * All declarations need to have some address.
 */
export interface BaseDeclaration extends BaseNode {
  datatype: TypeList
  name: string
  address: Address
}

/**
 * Possible forms of declarations.
 */
export type Declaration = VariableDeclaration | ArrayDeclaration

/**
 * Declaration of a variable.
 * E.g. `int x`, `int * x`, `struct Person p`.
 */
export interface VariableDeclaration extends BaseDeclaration {
  type: 'VariableDeclaration'
  init?: Expression | SequenceExpression | undefined
}

/**
 * Declaration of an array.
 * E.g. `int x[3]`, `int * x[3]`, `struct Person p[2]`.
 */
export interface ArrayDeclaration extends BaseDeclaration {
  type: 'ArrayDeclaration'
  init?: SequenceExpression | undefined
  size: number
}

/**
 * Declaration of a function.
 */
export interface FunctionDefinition extends BaseDeclaration {
  type: 'FunctionDefinition'
  params: Array<Identifier>
  body: Array<Statement>
}
