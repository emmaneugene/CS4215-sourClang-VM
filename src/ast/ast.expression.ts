import { Address, BaseNode, DataType, TypeList } from './ast.core'

/**
 * A base interface to describe expressions.
 * All expressions need to have some
 * datatype.
 */
export interface BaseExpression extends BaseNode {
  datatype: TypeList
}

/**
 * Describes a node defined using the following syntax:
 * `{ expr1, expr2, ... }`
 */
export interface SequenceExpression extends BaseExpression {
  type: 'SequenceExpression'
  expressions: Array<Expression>
}

/**
 * A union type of primary identifiers.
 * E.g. `x`, `list[1]`, `1`, `3.14`, `"hello"`
 */
export type PrimaryIdentifier = AddressableOperands | IntLiteral | FloatLiteral | StringLiteral

export interface IntLiteral extends BaseExpression {
  type: 'IntLiteral'
  value: number
}

export interface FloatLiteral extends BaseExpression {
  type: 'FloatLiteral'
  value: number
}

export interface StringLiteral extends BaseExpression {
  type: 'StringLiteral'
  datatype: {
    typeList: ['*', DataType.CHAR]
  }
  stringValue: string
  address: number
}

/**
 * A union type of addressable operands.
 * E.g. `x`, `list[1]`
 */
export type AddressableOperands = ArrayAccess | Identifier

export interface ArrayAccess extends BaseExpression {
  type: 'ArrayAccess'
  array: Identifier
  index: Expression
  address: Address
}

export interface Identifier extends BaseExpression {
  type: 'Identifier'
  name: string
  address: Address
}

/**
 * A union type of allowed expressions.
 */
export type Expression =
  | PrimaryIdentifier
  | UpdateExpression
  | FunctionCallExpression
  | NegationExpression
  | UnaryMinusExpression
  | CastExpression
  | DereferenceExpression
  | AddressOfExpression
  | SizeofExpression
  | BinaryOperatorExpression
  | LogicalExpression
  | TernaryExpression
  | SequenceExpression

export interface UpdateExpression extends BaseExpression {
  type: 'UpdateExpression'
  operand: AddressableOperands
  isPrefix: boolean
  operator: '++' | '--'
}

export interface FunctionCallExpression extends BaseExpression {
  type: 'FunctionCallExpression'
  callee: string
  arguments: Array<Expression>
}

export interface NegationExpression extends BaseExpression {
  type: 'NegationExpression'
  datatype: {
    typeList: [DataType.CHAR]
  }
  operand: Expression
}

export interface UnaryMinusExpression extends BaseExpression {
  type: 'UnaryMinusExpression'
  operand: Expression
}

export interface CastExpression extends BaseExpression {
  type: 'CastExpression'
  targetType: TypeList
  operand: Expression
}

export interface DereferenceExpression extends BaseExpression {
  type: 'DereferenceExpression'
  operand: Expression
}

export interface AddressOfExpression extends BaseExpression {
  type: 'AddressOfExpression'
  operand: AddressableOperands
}

export interface SizeofExpression extends BaseExpression {
  type: 'SizeofExpression'
  operand: AddressableOperands | TypeList
}

export interface BinaryOperatorExpression extends BaseExpression {
  type: 'BinaryOperatorExpression'
  operator: '+' | '-' | '*' | '/' | '%' | '>' | '>=' | '<' | '<=' | '==' | '!='
  left: Expression
  right: Expression
}

export interface LogicalExpression extends BaseExpression {
  type: 'LogicalExpression'
  operator: '||' | '&&'
  left: Expression
  right: Expression
}

export interface TernaryExpression extends BaseExpression {
  type: 'TernaryExpression'
  test: Expression
  alternate: Expression
  consequent: Expression
}
