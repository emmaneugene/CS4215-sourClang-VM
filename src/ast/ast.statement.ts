import { BaseNode } from './ast.core'
import { Declaration } from './ast.declaration'
import { AddressableOperands, Expression } from './ast.expression'

/**
 * A base interface to describe statements.
 */
export type BaseStatement = BaseNode

/**
 * A union type of statements.
 */
export type Statement =
  | ExpressionStatement
  | Declaration
  | CompoundStatement
  | AssignmentStatement
  | DerefLeftAssignmentStatement
  | IfElseStatement
  | WhileStatement
  | ForStatement
  | ReturnStatement
  | BreakStatement
  | ContinueStatement

export interface ExpressionStatement extends BaseStatement {
  type: 'ExpressionStatement'
  expression: Expression
}

export interface CompoundStatement extends BaseStatement {
  type: 'CompoundStatement'
  body: Array<Statement>
}

export interface AssignmentStatement extends BaseStatement {
  type: 'AssignmentStatement'
  operator: '='
  left: AddressableOperands
  right: Expression
}

export interface DerefLeftAssignmentStatement extends BaseStatement {
  type: 'DerefLeftAssignmentStatement'
  operator: '='
  derefChain: Array<'*'>
  left: AddressableOperands
  right: Expression
}

export interface IfElseStatement extends BaseStatement {
  type: 'IfElseStatement'
  test: Expression
  consequent: CompoundStatement
  alternate?: CompoundStatement | undefined
}

export interface WhileStatement extends BaseStatement {
  type: 'WhileStatement'
  test: Expression
  body: CompoundStatement
}

export interface ForStatement extends BaseStatement {
  type: 'ForStatement'
  init?: AssignmentStatement | DerefLeftAssignmentStatement | undefined
  test?: Expression | undefined
  update?: Expression | AssignmentStatement | DerefLeftAssignmentStatement | undefined
  body: CompoundStatement
}

export interface ReturnStatement extends BaseStatement {
  type: 'ReturnStatement'
  expression?: Expression | undefined
}

export interface BreakStatement extends BaseStatement {
  type: 'BreakStatement'
}

export interface ContinueStatement extends BaseStatement {
  type: 'ContinueStatement'
}
