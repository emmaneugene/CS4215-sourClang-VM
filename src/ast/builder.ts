import { WORD_SIZE } from '../constants'
import { getSizeofArrayVariable, getSizeofVariable } from '../utils/sizeHandler'
import { Program } from './ast.core'
import { ArrayDeclaration, FunctionDefinition, VariableDeclaration } from './ast.declaration'
import { Expression } from './ast.expression'
import { Statement, WhileStatement } from './ast.statement'
import { IdentifierHandler, IdentifierInfo } from './identifierHandler'

/**
 * Builds the AST.
 */
export class AstBuilder {
  // A mapping of the declared strings to their memory address
  // Initialised in the constructor
  private globalRODataMapping: Record<string, number> = {}

  // Used to track frames
  // Initialised in the constructor
  private identifierHandler: IdentifierHandler = new IdentifierHandler()

  // The root node of the AST being built
  private rootProgramNode: Program = {
    type: 'Program',
    body: []
  }

  /**
   * Constructs an AST builder instance.
   *
   * @param declaredStrings a list of strings in the user program.
   */
  constructor(declaredStrings: string[]) {
    this.initRODataMapping(declaredStrings)
  }

  /**
   * Creates a new function definition builder.
   *
   * It sets up the builder by giving it its own
   * function frame, but it doesn't place in
   * into the program node unless the builder's
   * build function is called.
   */
  createNewFunctionDefBuilder(args: Omit<FunctionDefinition, 'type'>): FunctionDefAstBuilder {
    const { name, params } = args
    this.identifierHandler.initFunctionFrame(name, params)
    const builder = new FunctionDefAstBuilder(
      args,
      this.identifierHandler,
      (node: FunctionDefinition) => {
        this.rootProgramNode.body.push(node)
      }
    )

    return builder
  }

  /**
   * Adds a global variable.
   */
  addGlobalVarDeclr(args: Omit<VariableDeclaration, 'type'>): VariableDeclaration {
    const { name, datatype } = args
    const varDeclaration: VariableDeclaration = {
      ...args,
      type: 'VariableDeclaration'
    }

    this.identifierHandler.addLocalVarToCurrentFrame({
      name,
      datatype,
      size: getSizeofVariable(varDeclaration)
    })

    this.rootProgramNode.body.push(varDeclaration)

    return varDeclaration
  }

  /**
   * Adds a global array variable.
   */
  addGlobalArrayDeclr(args: Omit<ArrayDeclaration, 'type'>): ArrayDeclaration {
    const { name, datatype } = args
    const arrayDeclaration: ArrayDeclaration = {
      ...args,
      type: 'ArrayDeclaration'
    }

    this.identifierHandler.addLocalVarToCurrentFrame({
      name,
      datatype,
      size: getSizeofArrayVariable(arrayDeclaration)
    })
    this.rootProgramNode.body.push(arrayDeclaration)

    return arrayDeclaration
  }

  /**
   * Initialise the address for RO data.
   *
   * @returns the total size of RO data segment
   */
  private initRODataMapping(list: string[]): number {
    let nextRODataAddr = 0
    list.forEach(s => {
      this.globalRODataMapping[s] = nextRODataAddr
      nextRODataAddr += (s.length + 1) * WORD_SIZE
    })
    return nextRODataAddr
  }
}

/**
 * Builds an AST subtree for a single function definition.
 */
export class FunctionDefAstBuilder {
  private functionDefinitionNode: FunctionDefinition
  private identifierHandler: IdentifierHandler
  private onBuildFunctionDef: (node: FunctionDefinition) => void

  /**
   * @param args the function info like name, return type and params
   * @param identifierHandler the handler for identifier
   * @param onBuildFunctionDef the function to call when the node has
   * been fully built
   */
  constructor(
    args: Omit<FunctionDefinition, 'type' | 'body'>,
    identifierHandler: IdentifierHandler,
    onBuildFunctionDef: (node: FunctionDefinition) => void
  ) {
    this.functionDefinitionNode = {
      ...args,
      type: 'FunctionDefinition',
      body: []
    }
    this.identifierHandler = identifierHandler
    this.onBuildFunctionDef = onBuildFunctionDef
  }

  addStatements(s: Statement[]): void {
    this.functionDefinitionNode.body.push(...s)
  }

  enterBlock(): void {
    this.identifierHandler.extendFrame()
  }

  exitBlock(): void {
    this.identifierHandler.popFrame()
  }

  /**
   * Creates a new expression builder.
   *
   * It sets up the builder by giving it a
   * lookup function to find variables that
   * it needs.
   */
  createNewExpressionBuilder(): ExpressionAstBuilder {
    return new ExpressionAstBuilder(this.identifierHandler.getIdentifierInfo)
  }

  build(): void {
    this.onBuildFunctionDef(this.functionDefinitionNode)
  }
}

/**
 * Builds an AST subtree for a single expression.
 */
export class ExpressionAstBuilder {
  private lookupFunction: (name: string) => IdentifierInfo

  /** The expression current being built. */
  private expressionProduct: Expression

  constructor(lookupFunction: (name: string) => IdentifierInfo) {
    this.lookupFunction = lookupFunction
  }

  buildExpression(): Expression {
    return this.expressionProduct
  }
}
