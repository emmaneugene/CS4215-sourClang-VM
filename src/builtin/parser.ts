import { DataType, TypeList } from '../ast/ast.core'
import { Microcode } from '../typings/microcode'

type BuiltInFunctionDefinition = {
  name: string
  datatype: TypeList
  instrs: Microcode[]
}

/**
 * Built-in functions.
 */
const PRINTF: BuiltInFunctionDefinition = {
  name: 'printf',
  datatype: {
    typeList: [DataType.LONG]
  },
  instrs: [{ type: 'ExecuteBuiltInFxCommand', name: 'printf' }, { type: 'ReturnCommand' }]
}

const MALLOC: BuiltInFunctionDefinition = {
  name: 'malloc',
  datatype: {
    typeList: ['*', DataType.VOID]
  },
  instrs: [{ type: 'ExecuteBuiltInFxCommand', name: 'malloc' }, { type: 'ReturnCommand' }]
}

const FREE: BuiltInFunctionDefinition = {
  name: 'free',
  datatype: {
    typeList: [DataType.VOID]
  },
  instrs: [{ type: 'ExecuteBuiltInFxCommand', name: 'free' }, { type: 'ReturnCommand' }]
}

export const BUILTINS: BuiltInFunctionDefinition[] = [PRINTF, MALLOC, FREE]
