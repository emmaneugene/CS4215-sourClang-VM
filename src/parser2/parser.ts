import { CharStreams, CommonTokenStream } from 'antlr4ts'

import { Program } from '../ast/ast.core'
import { SourCLexer } from '../lang/SourCLexer'
import { ProgramContext, SourCParser2 } from '../lang/SourCParser2'
import { FatalSyntaxError } from '../parser/error'
import { Context, ErrorSeverity } from '../types'
import { RODataSegment } from './rosegment'
import { ProgramGenerator } from './visitor.program'
import { RODataFinder } from './visitor.rodata'

export type ParseResult = {
  program: Program
  rodataSegment: RODataSegment
}

export function parse(source: string, context: Context): ParseResult | undefined {
  let parseResult: ParseResult | undefined

  const inputStream = CharStreams.fromString(source)
  const lexer = new SourCLexer(inputStream)
  const tokenStream = new CommonTokenStream(lexer)
  const parser = new SourCParser2(tokenStream)
  parser.buildParseTree = true

  try {
    const tree = parser.program()
    parseResult = buildAst(tree)
  } catch (error) {
    if (error instanceof FatalSyntaxError) {
      context.errors.push(error)
    } else {
      throw error
    }
  }

  const hasErrors = context.errors.find(m => m.severity === ErrorSeverity.ERROR)
  if (parseResult && !hasErrors) {
    return parseResult
  } else {
    return undefined
  }
}

function buildAst(ctx: ProgramContext): ParseResult {
  const rodataFinder = new RODataFinder()
  const rodataStrings = rodataFinder.visit(ctx)
  const rodataSegment = new RODataSegment(rodataStrings)

  const tokenTreeVisitor = new ProgramGenerator(rodataSegment.getRODataSegmentSize())
  tokenTreeVisitor.visitProgram(ctx)
  const program = tokenTreeVisitor.getAst()
  return {
    program,
    rodataSegment
  }
}
