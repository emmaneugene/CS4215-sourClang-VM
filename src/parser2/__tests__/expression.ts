import { VariableDeclaration } from '../../ast/ast.declaration'
import createContext from '../../createContext'
import { Context, Variant } from '../../types'
import { parse, ParseResult } from '../parser'

const f = (when: string, should: string) => `When_\'${when}\'_Should_${should}`
const prependFixture = (expr: string) => `int x = 1; int y = ${expr};`
const getBinop = (result: ParseResult | undefined) => {
  if (!result) {
    return undefined
  }
  const v = result?.program.body[0] as VariableDeclaration
  const stmt = v.init
  return stmt
}

let context: Context = createContext(Variant.DEFAULT, undefined, undefined)

beforeAll(() => {
  console.log = () => {}
})

beforeEach(() => {
  context = createContext(Variant.DEFAULT, undefined, undefined)
})

const ADD_TEST_CASES: string[] = ['1+2', 'x+5']

describe('Addition', () => {
  for (const source of ADD_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(prependFixture(source), context)
      const binop = getBinop(result)
      expect(binop).toBeTruthy()
      expect(binop).toMatchSnapshot()
    })
  }
})

const SUB_TEST_CASES: string[] = ['1-2', 'x-5', '-x-5']

describe('Subtraction', () => {
  for (const source of SUB_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(prependFixture(source), context)
      const binop = getBinop(result)
      expect(binop).toBeTruthy()
      expect(binop).toMatchSnapshot()
    })
  }
})

const MULT_TEST_CASES: string[] = ['1*2', 'x*5']

describe('Multiplication', () => {
  for (const source of MULT_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(prependFixture(source), context)
      const binop = getBinop(result)
      expect(binop).toBeTruthy()
      expect(binop).toMatchSnapshot()
    })
  }
})

const DIV_TEST_CASES: string[] = ['1/2', 'x/5']

describe('Division', () => {
  for (const source of DIV_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(prependFixture(source), context)
      const binop = getBinop(result)
      expect(binop).toBeTruthy()
      expect(binop).toMatchSnapshot()
    })
  }
})

const OPERAND3_TEST_CASES: string[] = ['1/2 + 4000', 'x/5 + 0 * (-x)', 'x/1 + (-1) * ((-1) * x)']

describe('3Operands', () => {
  for (const source of OPERAND3_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(prependFixture(source), context)
      const binop = getBinop(result)
      expect(binop).toBeTruthy()
      expect(binop).toMatchSnapshot()
    })
  }
})
