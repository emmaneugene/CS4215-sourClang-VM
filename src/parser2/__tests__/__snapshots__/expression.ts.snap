// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3Operands When_'1/2 + 4000'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`3Operands When_'x/1 + (-1) * ((-1) * x)'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`3Operands When_'x/5 + 0 * (-x)'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Addition When_'1+2'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Addition When_'x+5'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Division When_'1/2'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Division When_'x/5'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Multiplication When_'1*2'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Multiplication When_'x*5'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Subtraction When_'-x-5'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Subtraction When_'1-2'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;

exports[`Subtraction When_'x-5'_Should_ParseSuccessfully 1`] = `
Object {
  "datatype": Object {
    "typeList": Array [
      "long",
    ],
  },
  "end": Object {
    "column": 8,
    "line": 1,
  },
  "start": Object {
    "column": 8,
    "line": 1,
  },
  "type": "IntLiteral",
  "value": 1,
}
`;
